src/lib.rs

```diff
@@ -362,204 +333,173 @@ where
-                Ok(1)
-            } else {
-                Err(io::Error::from(io::ErrorKind::WouldBlock))
-            }
-        }
-    }
-
     impl AsyncRead for DelayStream {
-        unsafe fn prepare_uninitialized_buffer(&self, _: &mut [u8]) -> bool {
-            true
+        fn poll_read(
+            self: Pin<&mut Self>,
+            cx: &mut Context,
+            _buf: &mut [u8],
+        ) -> Poll<Result<usize, io::Error>> {
+            let s = self.get_mut();
+            match Pin::new(&mut s.0).poll(cx) {
+                Poll::Ready(()) => Poll::Ready(Ok(1)),
+                Poll::Pending => Poll::Pending,
+            }
         }
     }
 
-    impl Write for DelayStream {
-        fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-            if self.0
-                .poll()
-                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?
-                .is_ready()
-            {
-                Ok(buf.len())
-            } else {
-                Err(io::Error::from(io::ErrorKind::WouldBlock))
-            }
+    impl AsyncWrite for DelayStream {
+        fn poll_shutdown(
+            self: Pin<&mut Self>,
+            _cx: &mut Context,
+        ) -> Poll<Result<(), io::Error>> {
+            Poll::Ready(Ok(()))
         }
 
-        fn flush(&mut self) -> io::Result<()> {
-            Ok(())
+        fn poll_flush(
+            self: Pin<&mut Self>,
+            _cx: &mut Context,
+        ) -> Poll<Result<(), io::Error>> {
+            Poll::Ready(Ok(()))
         }
-    }
 
-    impl AsyncWrite for DelayStream {
-        fn shutdown(&mut self) -> Poll<(), io::Error> {
-            Ok(Async::Ready(()))
+        fn poll_write(
+            self: Pin<&mut Self>,
+            cx: &mut Context,
+            buf: &[u8],
+        ) -> Poll<Result<usize, io::Error>> {
+            let s = self.get_mut();
+            match Pin::new(&mut s.0).poll(cx) {
+                Poll::Ready(()) => Poll::Ready(Ok(buf.len())),
+                Poll::Pending => Poll::Pending,
+            }
         }
     }
 
-    struct ReadFuture<S>(Option<S>);
-
-    impl<S> Future for ReadFuture<S>
+    async fn read_one<R>(mut reader: R) -> Result<usize, io::Error>
     where
-        S: AsyncRead,
+        R: AsyncRead + Unpin,
     {
-        type Item = S;
-        type Error = io::Error;
-
-        fn poll(&mut self) -> Poll<S, io::Error> {
-            let mut reader = self.0.take().unwrap();
-
-            let mut buf = [0; 1];
-            match reader.read(&mut buf) {
-                Ok(_) => Ok(Async::Ready(reader)),
-                Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
-                    self.0 = Some(reader);
-                    Ok(Async::NotReady)
-                }
-                Err(e) => Err(e),
-            }
-        }
+        let mut buf: [u8; 1] = [0; 1];
+        reader.read(&mut buf).await
     }
 
-    fn run<F>(f: F) -> Result<F::Item, F::Error>
+    fn run<F>(f: F) -> F::Output
```

Comment:
All of the tests can switch to using `#[tokio::test]` to avoid having to manually manage runtimes and such.

Reply:
I'll look into moving them over tonight.